File: ios/Jamboree/Tutorials/TutorialVote.swift
import Foundation

/// Pure majority-wins yes/no vote used to decide whether a game's pre-game
/// tutorial should be shown. Engines and servers wrap this; it has no I/O.
final class TutorialVote {
    private(set) var isOpen = false
    private var votes: [String: Bool] = [:]
    private(set) var result: Bool?
    /// Tutorial was actually displayed to players — used to keep the vote
    /// button hidden after the tutorial has already run once.
    private(set) var tutorialShown = false
    private(set) var eligible: Set<String> = []

    var hasResult: Bool { result != nil }
    var yesCount: Int { votes.values.filter { $0 }.count }
    var noCount: Int { votes.values.filter { !$0 }.count }
    var eligibleCount: Int { eligible.count }

    func markShown() { tutorialShown = true }
    func isEligible(_ id: String) -> Bool { eligible.contains(id) }

    /// Reset and open a fresh vote. `eligibleIds` is the snapshot of players
    /// in the lobby when the vote was called.
    func open<S: Sequence>(eligibleIds: S) where S.Element == String {
        isOpen = true
        votes.removeAll()
        result = nil
        eligible = Set(eligibleIds)
    }

    /// Returns true once every eligible voter has submitted — at which point
    /// [result] is finalized.
    @discardableResult
    func submit(voterId: String, yes: Bool) -> Bool {
        guard isOpen, eligible.contains(voterId) else { return false }
        votes[voterId] = yes
        if votes.count >= eligible.count {
            finalize()
            return true
        }
        return false
    }

    func close() {
        guard isOpen else { return }
        finalize()
    }

    /// A voter dropped out (e.g. left the lobby).
    func removeVoter(_ id: String) {
        guard isOpen else { return }
        eligible.remove(id)
        votes[id] = nil
        if eligible.isEmpty {
            isOpen = false
            votes.removeAll()
            result = nil
            return
        }
        if votes.count >= eligible.count { finalize() }
    }

    private func finalize() {
        // Strict majority — ties resolve to "no" (skip tutorial).
        result = yesCount > noCount
        isOpen = false
    }
}

File: android/app/src/main/java/com/example/jamboree/tutorials/TutorialVote.kt
package com.example.ubapp.tutorials

/**
 * Pure majority-wins yes/no vote used to decide whether a game's pre-game
 * tutorial should be shown. Engines and servers wrap this; it has no I/O.
 */
class TutorialVote {
    var isOpen: Boolean = false; private set
    private val votes: MutableMap<String, Boolean> = mutableMapOf()
    var result: Boolean? = null; private set
    /** Tutorial was actually displayed to players — used to keep the vote
     *  button hidden after the tutorial has already run once. */
    var tutorialShown: Boolean = false; private set
    private val eligibleSet: MutableSet<String> = mutableSetOf()

    val hasResult: Boolean get() = result != null
    val yesCount: Int get() = votes.values.count { it }
    val noCount: Int get() = votes.values.count { !it }
    val eligibleCount: Int get() = eligibleSet.size
    val eligible: Set<String> get() = eligibleSet

    fun markShown() { tutorialShown = true }
    fun isEligible(id: String) = eligibleSet.contains(id)

    /** Reset and open a fresh vote. [eligibleIds] is the snapshot of players
     *  in the lobby when the vote was called. */
    fun open(eligibleIds: Iterable<String>) {
        isOpen = true
        votes.clear()
        result = null
        eligibleSet.clear()
        eligibleSet.addAll(eligibleIds)
    }

    /** Returns true once every eligible voter has submitted — at which point
     *  [result] is finalized. */
    fun submit(voterId: String, yes: Boolean): Boolean {
        if (!isOpen || voterId !in eligibleSet) return false
        votes[voterId] = yes
        if (votes.size >= eligibleSet.size) { finalize(); return true }
        return false
    }

    fun close() { if (isOpen) finalize() }

    /** A voter dropped out (e.g. left the lobby). */
    fun removeVoter(id: String) {
        if (!isOpen) return
        eligibleSet.remove(id)
        votes.remove(id)
        if (eligibleSet.isEmpty()) {
            isOpen = false; votes.clear(); result = null
            return
        }
        if (votes.size >= eligibleSet.size) finalize()
    }

    private fun finalize() {
        // Strict majority — ties resolve to "no" (skip tutorial).
        result = yesCount > noCount
        isOpen = false
    }
}
